lib/srv/desktop/rdp/rdpclient/src/lib.rs

```diff
@@ -0,0 +1,359 @@
+    let tcp = match TcpStream::connect(&addr) {
+        Ok(tcp) => tcp,
+        Err(e) => return format!("failed TCP connection to {}: {:?}", &addr, e).into(),
+    };
+    let tcp_fd = tcp.as_raw_fd() as usize;
+    let mut connector = Connector::new()
+        .screen(screen_width, screen_height)
+        .credentials(".".to_string(), username.to_string(), password.to_string());
+    let client = match connector.connect(tcp) {
+        Ok(client) => client,
+        Err(e) => return format!("failed RDP connection to {}: {:?}", &addr, e).into(),
+    };
+
+    register_client(
+        client_ref,
+        Client {
+            rdp_client: client,
+            tcp_fd: tcp_fd,
+        },
+    );
+    CGOError::null()
+}
+
+#[repr(C)]
+pub struct CGOBitmap {
+    pub dest_left: u16,
+    pub dest_top: u16,
+    pub dest_right: u16,
+    pub dest_bottom: u16,
+    // Memory is freed on the Rust side.
+    pub data_ptr: *const u8,
+    pub data_len: usize,
+}
+
+impl TryFrom<BitmapEvent> for CGOBitmap {
+    type Error = Error;
+
+    fn try_from(e: BitmapEvent) -> Result<Self, Self::Error> {
+        let mut res = CGOBitmap {
+            dest_left: e.dest_left,
+            dest_top: e.dest_top,
+            dest_right: e.dest_right,
+            dest_bottom: e.dest_bottom,
+            data_ptr: std::ptr::null(),
+            data_len: 0,
+        };
+
+        // e.decompress consumes e, so we need to call it separately, after populating the fields
+        // above.
+        let data = if e.is_compress {
+            e.decompress()?
+        } else {
+            e.data
+        };
+        res.data_ptr = data.as_ptr();
+        res.data_len = data.len();
+
+        Ok(res)
+    }
+}
+
+// TODO: this is Linux-specific, also implement for Windows.
+fn wait_for_fd(fd: usize) -> bool {
+    unsafe {
+        let mut raw_fds: fd_set = mem::zeroed();
+
+        FD_SET(fd as i32, &mut raw_fds);
+
+        let result = select(
+            fd as i32 + 1,
+            &mut raw_fds,
+            ptr::null_mut(),
+            ptr::null_mut(),
+            ptr::null_mut(),
+        );
+        result == 1
+    }
+}
+
+// read_rdp_output reads incoming RDP bitmap frames from client at client_ref and forwards them to
+// handle_bitmap. handle_bitmap *must not* free the memory of CGOBitmap.
+#[no_mangle]
+pub extern "C" fn read_rdp_output(
+    client_ref: i64,
+    handle_bitmap: unsafe extern "C" fn(i64, CGOBitmap) -> CGOError,
+) -> CGOError {
+    let mut tcp_fd = 0;
+    if let Err(e) = with_client(&client_ref, |client| -> Result<(), ClientError> {
+        tcp_fd = client.lock().unwrap().tcp_fd;
+        Ok(())
+    }) {
+        return format!(
+            "failed looking up TCP file descriptor for client {}: {:?}",
+            client_ref, e
+        )
+        .into();
+    }
+    // Read incoming events.
+    // TODO: this doesn't always unblock after client.shutdown() was called. Figure out why.
+    while wait_for_fd(tcp_fd as usize) {
```

Comment:
This uh seems like a loop that would gobble a lot of CPU as it's constantly making syscalls. Can you passively wait instead or is this not a problem?

Reply:
ah, clever 👍